service/src/event/validator/signed.rs

```diff
@@ -0,0 +1,252 @@
+                                }
+                            }
+                        } else {
+                            None
+                        };
+
+                        match self.verify_signature(init, signed, opts).await {
+                            Ok(_) => {
+                                self.add_valid(event);
+                            }
+                            Err(err) => self.add_invalid(InvalidItem::InvalidSignature {
+                                key: event.order_key().to_owned(),
+                                reason: err,
+                            }),
+                        }
+                    }
+                    unvalidated::Payload::Init(init) => {
+                        // yes, we are the init in this case but it still works
+                        match self.verify_signature_against_init(init, signed, opts).await {
+                            Ok(_) => {
+                                self.track_init(*event.cid(), event.event().to_owned());
+                                self.add_valid(event)
+                            }
+                            Err(err) => {
+                                self.invalid_init.insert(*event.cid());
+                                self.add_invalid(InvalidItem::InvalidSignature {
+                                    key: event.order_key().to_owned(),
+                                    reason: err,
+                                })
+                            }
+                        }
+                    }
+                },
+                unvalidated::Event::Unsigned(_) => {
+                    unreachable!("validate_signed does not see unsigned events")
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn verify_signature(
+        &mut self,
+        init: Option<Arc<unvalidated::Event<Ipld>>>,
+        signed: &signed::Event<Ipld>,
+        opts: &VerifyJwsOpts,
+    ) -> std::result::Result<(), String> {
+        match init.as_deref() {
+            Some(unvalidated::Event::Time(_)) => {
+                Err("init event pointer cannot reference a time event".to_string())
+            }
+            Some(unvalidated::Event::Signed(s)) => match s.payload() {
+                unvalidated::Payload::Data(_) => {
+                    Err("init event pointer cannot reference a data event".to_string())
+                }
+                unvalidated::Payload::Init(init) => {
+                    self.verify_signature_against_init(init, signed, opts).await
+                }
+            },
+            Some(unvalidated::Event::Unsigned(init)) => {
+                self.verify_signature_against_init(init, signed, opts).await
+            }
+            None => {
+                // TODO: we skip validating the CACAO resources without the init because we don't even know
+                // what stream we're on, so it will fail if we don't have * resources access and the data
```

Comment:
this should actually say what model (sepKey/sepValue) rather than stream.. we can assume our `id` CID is the real stream ID but we may not be able to pass the time checks so I still think we should skip checking the cacao resources